### ISSUE TITLE
[Question] Interface with other UI packages

### ISSUE THREAD

**Issue by GDelevoye** [QuantumBadger/Speedy2D#25]
Hi,

Thank you for your package, which looks exactly what I was looking for, in terms of simplicity and code design

I'm a somewhat experienced programmer, but I'm quite new in rust and also, very new to graphical programming. So my question will be really naive : Can I create widgets with another library (e.g druid, conrod fltk-rs etc), and then link them to a window created with Speedy2D ? And if yes, do you know if there is one extern library in particular that could be easier to use with Speedy2D ? Preferably in full rust ?

I don't need sophisticated widgets (see picture)

![image-asset](https://user-images.githubusercontent.com/29506993/115883822-8292df80-a44e-11eb-9016-f6a558890eaf.jpeg)

**Comment by wyhinton** [comment]
Yes you can. I didn't add any buttons in the "[Speedy2d Canvas](https://github.com/wyhinton/FLTK-RS-Examples)" example, but it shows using fltk-rs with speedy2d and adding buttons is pretty trivial. Any UI library that provides a gl context should also work.

**Comment by QuantumBadger** [comment]
@GDelevoye Thanks for the questions! Speedy2D needs a dedicated OpenGL context that isn't used by any other libraries. This is because a context contains global configuration, so multiple libraries using the same context would conflict with each other.

There are a few ways to combine Speedy2D with a UI library:

* Use a library that lets you create new OpenGL rendering surfaces within a window, for example Qt
* Modify a library to use Speedy2D APIs for rendering text/shapes/images, so that the UI library doesn't need direct access to OpenGL
* Create a context for Speedy2D which renders to a texture, and then render that texture in the UI library

Each thread in a program can have one OpenGL context "globally active" at a time, so if you're using multiple contexts in a single thread, you'd need to manually switch between them.

To make use of egui, it sounds like you'd need to create an integration:

https://github.com/emilk/egui#integrations

This would mean that draw commands from egui were sent to Speedy2D, and input events from Speedy2D were sent to egui.

**Comment by GDelevoye** [comment]
Ok ! I think it's much clearer now

I'll give a try like this, and maybe I'll come back to you if I have other questions but it looks like a cool start :) Thank you very much
